File: frontend/lib/documento.ts
/**
 * Qué documento es, por su largo. Es la única regla que distingue a una
 * persona natural de una empresa en todo el ERP de cara al cliente: 8 dígitos
 * es DNI (RENIEC, boleta), 11 es RUC (SUNAT, factura) — RN-CPP-003.
 *
 * Vive acá y no dentro de un componente por lo mismo que `permisos.ts`:
 * `npm test` corre sobre Node pelado, que hace type-stripping pero no
 * transforma JSX. La regla que decide a qué padrón se le pregunta tiene que
 * poder probarse sin montar React.
 */

export const LARGO_DNI = 8;
export const LARGO_RUC = 11;

export type TipoConsulta = "dni" | "ruc";

/**
 * `null` cuando el número todavía no es ninguno de los dos: 7 dígitos no es
 * "un DNI casi listo", es un número que no se puede consultar. Se prefiere
 * decirlo a adivinar — una consulta a ciegas gasta cuota de un proveedor
 * pago y vuelve con "no encontrado" que no significa nada.
 */
export function tipoPorLargo(numero: string): TipoConsulta | null {
  const limpio = numero.trim();
  if (!/^\d+$/.test(limpio)) return null;
  if (limpio.length === LARGO_DNI) return "dni";
  if (limpio.length === LARGO_RUC) return "ruc";
  return null;
}

/** Vacío también vale: se vende sin documento (RN-PER-005), y el
 * comprobante sale a nombre de «Clientes varios». */
export function documentoValido(numero: string): boolean {
  return numero.length === 0 || tipoPorLargo(numero) !== null;
}

/**
 * El vocabulario de `persona.tipo_documento`, espejo de `src/shared/documento.py`.
 * Estaba escrito a mano en tres formularios que no coincidían —uno ofrecía
 * `ruc`, otro `carne_extranjeria`— y el backend rechazaba los dos con un 500.
 */
export const TIPOS_DOCUMENTO = ["dni", "ce", "pasaporte", "ruc"] as const;

export type TipoDocumento = (typeof TIPOS_DOCUMENTO)[number];

/** Los de una persona natural: en Ventas, 11 dígitos hacen al cliente
 * jurídico y su RUC va en el cliente, no en la persona. */
export const TIPOS_DOCUMENTO_NATURAL = ["dni", "ce", "pasaporte"] as const;

export const ETIQUETA_DOCUMENTO: Record<TipoDocumento, string> = {
  dni: "DNI",
  ce: "Carné de extranjería",
  pasaporte: "Pasaporte",
  ruc: "RUC",
};

/** Los que son un número de largo fijo. CE y pasaporte no: los emite otro país. */
export const LARGO_EXACTO: Partial<Record<TipoDocumento, number>> = {
  dni: LARGO_DNI,
  ruc: LARGO_RUC,
};

export const LARGO_MAXIMO_DOCUMENTO = 20;
const LARGO_MINIMO_DOCUMENTO = 6;

/** Vacío vale: el documento es opcional (ADR-018). Con algo escrito, el largo
 * lo manda el tipo — es la misma regla que `shared.documento.validar`. */
export function documentoValidoPorTipo(tipo: string, numero: string): boolean {
  const limpio = numero.trim();
  if (limpio.length === 0) return true;
  const exacto = LARGO_EXACTO[tipo as TipoDocumento];
  if (exacto !== undefined) return /^\d+$/.test(limpio) && limpio.length === exacto;
  return (
    /^[a-zA-Z0-9]+$/.test(limpio) &&
    limpio.length >= LARGO_MINIMO_DOCUMENTO &&
    limpio.length <= LARGO_MAXIMO_DOCUMENTO
  );
}

File: src/shared/documento.py
"""Tipo y número de documento de identidad: el vocabulario único del ERP.

Estaba escrito en cuatro sitios que no coincidían —el `Enum` de `persona`, la
tupla de `sales.application.clientes`, y dos listas de `<option>` en el
frontend—, y elegir «RUC» en el alta de persona devolvía 500: SQLAlchemy
rechaza en Python el valor que el formulario ofrecía. Un vocabulario partido no
falla el día que se escribe, falla el día que alguien usa la opción que sobra.

Vive en `shared` y no en `sales/domain/rules.py` porque `users` y `rrhh` no
pueden importar el dominio de otro módulo (CLAUDE.md); `rules` reexporta los
largos desde acá para no volver a tener dos copias.

**RUC de persona natural**: en Perú una persona natural con negocio tiene RUC
(empieza en 10), así que es un tipo legítimo de `persona` y no solo de empresa.
Distinto del `ruc` de `cliente`/`proveedor`/`empresa`, que identifica a una
persona jurídica y tiene columna propia.
"""

DNI = "dni"
CE = "ce"
PASAPORTE = "pasaporte"
RUC = "ruc"

#: Todo lo que `persona.tipo_documento` acepta.
TIPOS = (DNI, CE, PASAPORTE, RUC)

#: Los de una persona natural sin negocio. `sales` distingue: un documento de
#: 11 dígitos hace al cliente **jurídico**, y ese RUC no va en la persona.
NATURALES = (DNI, CE, PASAPORTE)

#: Lo que llega de fuera con otro nombre. El tablero de contratación mandaba
#: `carne_extranjeria` desde que existe, y romper ese formulario para que diga
#: `ce` no le sirve a nadie: se normaliza al entrar.
ALIAS = {
    "carne_extranjeria": CE,
    "carnet_extranjeria": CE,
    "carné_extranjería": CE,
}

LARGO_DNI = 8
LARGO_RUC = 11

#: Los que son un número de largo fijo. CE y pasaporte no: los emite otro país.
LARGO_EXACTO = {DNI: LARGO_DNI, RUC: LARGO_RUC}

#: El de la columna `persona.numero_documento`.
LARGO_MAXIMO = 20
LARGO_MINIMO = 6

ETIQUETA = {
    DNI: "DNI",
    CE: "carné de extranjería",
    PASAPORTE: "pasaporte",
    RUC: "RUC",
}


def normalizar(tipo: str | None) -> str | None:
    """Minúsculas, sin espacios y con los alias resueltos. `None` si viene
    vacío — el documento es opcional desde ADR-018."""
    if tipo is None:
        return None
    limpio = tipo.strip().lower()
    if not limpio:
        return None
    return ALIAS.get(limpio, limpio)


def validar(tipo: str | None, numero: str | None) -> tuple[str | None, str | None]:
    """Devuelve el par ya normalizado, o levanta `ValueError`.

    Sin documento es válido: un cliente de mostrador no siempre lo da
    (ADR-018). Con número **sí** hace falta el tipo, porque es lo que decide
    a qué padrón se consulta y qué se declara a SUNAT.
    """
    tipo = normalizar(tipo)
    numero = (numero or "").strip() or None
    if tipo is not None and tipo not in TIPOS:
        raise ValueError(
            f"tipo de documento inválido: «{tipo}». Debe ser uno de: {', '.join(TIPOS)}"
        )
    if numero is None:
        return tipo, None
    if tipo is None:
        raise ValueError("un número de documento necesita decir de qué tipo es")

    exacto = LARGO_EXACTO.get(tipo)
    if exacto is not None:
        if not numero.isdigit() or len(numero) != exacto:
            raise ValueError(f"un {ETIQUETA[tipo]} son {exacto} dígitos")
    elif not numero.isalnum() or not LARGO_MINIMO <= len(numero) <= LARGO_MAXIMO:
        raise ValueError(
            f"un {ETIQUETA[tipo]} son de {LARGO_MINIMO} a {LARGO_MAXIMO} "
            "caracteres alfanuméricos"
        )
    return tipo, numero
